src/v3/src/util/widgetHooks.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import { IdxTransaction } from '@okta/okta-auth-js';
+
+import { HookFunction, HooksOptions, HookType } from '../../../types';
+import { getFormNameForTransaction } from './getEventContext';
+
+export class WidgetHooks {
+  private hooks: Map<string, Map<string, HookFunction[]>>;
+
+  /* eslint-disable no-restricted-syntax */
+  constructor(hooksOptions?: HooksOptions) {
+    this.hooks = new Map();
+    if (hooksOptions) {
+      for (const [formName, formHooks] of Object.entries(hooksOptions)) {
+        for (const [hookType, hooksByType] of Object.entries(formHooks)) {
+          for (const hook of (hooksByType || []) as HookFunction[]) {
+            this.addHook(hookType as HookType, formName, hook);
+          }
+        }
```

Comment:
TS is making you cast `hookType as HookType` because it can't infer the types of anything that isn't `before` or `after` so it gives you `hooksByType: any`. We both know there will be only `before` or `after` but TS isn't smart enough to know that. Since there's only two cases we could avoid the type cast by directly handling both cases explicitly and writing the inner `for` loop write, once for each case.

Another option is to force TS to understand by using a string array of the literals we care about, which effectively does the same thing but explicitly:
```ts
['before', 'after'].forEach(hookType => {
  for (const hook of (formHooks[hookType] || []) as HookFunction[]) {
    this.addHook(hookType, formName, hook);
  }
});
```